pkg/server/plugin/noderesolver/azure/msi.go

```diff
@@ -0,0 +1,312 @@
+				return nil, msiError.Wrap(err)
+			}
+			tenantClients[tenantID] = p.hooks.newClient(tenant.SubscriptionId, authorizer)
+		}
+	}
+
+	p.setClients(msiClient, tenantClients)
+	return &spi.ConfigureResponse{}, nil
+}
+
+func (p *MSIResolverPlugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+func (p *MSIResolverPlugin) getClient(tenantID string) (apiClient, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	switch {
+	case p.msiClient != nil:
+		return p.msiClient, nil
+	case p.tenantClients != nil:
+		client, ok := p.tenantClients[tenantID]
+		if !ok {
+			return nil, msiError.New("not configured for tenant %q", tenantID)
+		}
+		return client, nil
+	default:
+		return nil, msiError.New("not configured")
+	}
+}
+
+func (p *MSIResolverPlugin) setClients(msiClient apiClient, tenantClients map[string]apiClient) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.msiClient = msiClient
+	p.tenantClients = tenantClients
+}
+
+func (p *MSIResolverPlugin) resolveSpiffeID(ctx context.Context, spiffeID string) (*common.Selectors, error) {
+	// parse out the tenant ID and principal ID from the token
+	u, err := idutil.ParseSpiffeID(spiffeID, idutil.AllowAnyTrustDomainAgent())
+	if err != nil {
+		return nil, msiError.Wrap(err)
+	}
+
+	tenantID, principalID, err := parseAgentIDPath(u.Path)
+	if err != nil {
+		logrus.Warnf("unrecognized Agent ID %q", spiffeID)
+		return nil, nil
+	}
+
+	client, err := p.getClient(tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Retrieve the resource belonging to the principal id.
+	vmResourceID, err := client.GetVirtualMachineResourceID(ctx, principalID)
+	if err != nil {
+		return nil, msiError.New("unable to get resource for principal %q: %v", principalID, err)
+	}
+
+	// parse out the resource group and vm name from the resource ID
+	vmResourceGroup, vmName, err := parseVirtualMachineID(vmResourceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// build up a unique map of selectors. this is easier than deduping
+	// individual selectors (e.g. the virtual network for each interface)
+	selectorMap := make(map[string]bool)
+	addSelector := func(parts ...string) {
+		selectorMap[strings.Join(parts, ":")] = true
+	}
+
+	addSelector("subscription-id", client.SubscriptionID())
+	addSelector("vm-name", vmResourceGroup, vmName)
+
+	// pull the VM information
+	vm, err := client.GetVirtualMachine(ctx, vmResourceGroup, vmName)
+	if err != nil {
+		return nil, msiError.New("unable to get virtual machine %q: %v", resourceGroupName(vmResourceGroup, vmName), err)
+	}
+
+	if np := vm.NetworkProfile; np != nil {
+		if nis := np.NetworkInterfaces; nis != nil {
+			for _, ni := range *nis {
+				if ni.ID == nil {
+					continue
+				}
+				niResourceGroup, niName, err := parseNetworkInterfaceID(*ni.ID)
+				if err != nil {
+					return nil, err
+				}
+				networkInterface, err := client.GetNetworkInterface(ctx, niResourceGroup, niName)
```

Comment:
how is this `networkInterface` var different from `ni` which also seems to stand for network interface?

Reply:
`NetworkInterfaces` is a slice of `NetworkInterfaceReference`'s. I will change the name to be more clear.